File: androidApp/src/main/java/com/nablet/mytasks/android/presentation/task_list/TasksViewModel.kt
package com.nablet.mytasks.android.presentation.task_list

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.nablet.mytasks.android.BuildConfig
import com.nablet.mytasks.domain.model.GenericMessageInfo
import com.nablet.mytasks.domain.util.*
import com.nablet.mytasks.presentation.task_list.TaskListEvent
import com.nablet.mytasks.presentation.task_list.TaskListState
import com.nablet.mytasks.usecases.*
import com.nablet.mytasks.util.Logger
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class TasksViewModel @Inject constructor(
	private val checkVersionUseCase: CheckVersion,
	private val loadTasksUseCase: LoadTasks,
	private val addTaskUseCase: AddTask,
	private val deleteTaskUseCase: DeleteTask,
) : ViewModel() {

	private val logger = Logger("TasksViewModel")

	private val _state = MutableStateFlow(TaskListState())
	val state: StateFlow<TaskListState>
		get() = _state.asStateFlow()

	fun onEvent(event: TaskListEvent) {
		when (event) {
			is TaskListEvent.AddTask -> addTask(event.name, event.desc)
			is TaskListEvent.OnSwipeRight -> deleteTask(event.selectedTask.name)
			is TaskListEvent.OnSwipeLeft -> deleteTask(event.selectedTask.name)
			is TaskListEvent.OnClick -> {}
			is TaskListEvent.OnLongClick -> {}
			is TaskListEvent.OnRemoveHeadMessageFromQueue -> removeHeadMessage()
		}
	}

	init {
		loadTasks()
		checkVersion()
	}

	private fun checkVersion() {
		checkVersionUseCase.execute(BuildConfig.VERSION_NAME)
			.collectCommon(viewModelScope) { status ->
				when (status) {
					is CheckVersionStatus.Loading -> _state.update { it.copy(isLoading = status.loading) }
					is CheckVersionStatus.Error -> appendToMessageQueue(status.message)
					is CheckVersionStatus.IsLatestVersion -> {}
					is CheckVersionStatus.NewVersionAvailable -> appendToMessageQueue(status.message)
				}
			}
	}

	private fun loadTasks() {
		loadTasksUseCase.execute().collectCommon(viewModelScope) { status ->
			when (status) {
				is LoadTasksOutput.Loading -> _state.update { it.copy(isLoading = status.loading) }
				is LoadTasksOutput.Data -> _state.update { it.copy(tasks = status.tasks) }
				is LoadTasksOutput.Error -> appendToMessageQueue(status.message)
			}
		}
	}

	private fun addTask(name: String, desc: String) {
		addTaskUseCase.execute(name, desc)?.also { appendToMessageQueue(it) }
		loadTasks()
	}

	private fun deleteTask(taskName: String) {
		deleteTaskUseCase.execute(taskName)?.also { appendToMessageQueue(it) }
		loadTasks()
	}

	private fun removeHeadMessage() = viewModelScope.launch {
		runCatching {
			_state.update { it.copy(queue = it.queue.apply { remove() }) }
		}.onFailure {
			logger.log("removeHeadMessage(): ${it.message}")
			_state.value = _state.value.copy(queue = Queue(mutableListOf()))
		}
	}

	private fun appendToMessageQueue(message: GenericMessageInfo.Builder) {
		val alreadyExistsInQueue = GenericMessageInfoQueueUtil().doesMessageAlreadyExistInQueue(
			queue = state.value.queue,
			messageInfo = message.build()
		)
		if (!alreadyExistsInQueue) {
			_state.update {
				it.copy(queue = it.queue.also { queue ->
					queue.add(message.build())
				})
			}
		}
	}

}

File: iosApp/iosApp/presentation/task_list/TasksViewModel.swift
//
//  TasksViewModel.swift
//  iosApp
//
//  Created by Jayson Tiongco on 1/22/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import SwiftUI
import shared

class TasksViewModel : ObservableObject {
    
    private let logger = Logger(className: "TasksViewModel")

    // UseCases
    let addTaskUseCase: AddTask
    let deleteTaskUseCase: DeleteTask
    let loadTasksUseCase: LoadTasks

    // State
    @Published var state: TaskListState = TaskListState()
    @Published var showDialog: Bool = false
    @Published var showAddTaskDialog: Bool = false

    init (usecasesModule: UseCasesModule) {
        self.addTaskUseCase = usecasesModule.addTask
        self.deleteTaskUseCase = usecasesModule.deleteTask
        self.loadTasksUseCase = usecasesModule.loadTasks
        
        for i in 1...20 {
            addTask(name: "Task name \(i)", desc: "Task description \(i)")
        }
        
        loadTasks()
    }
    
    func doNothing() {}
    
    func onUserEvent(userEvent: TaskListEvent) {
        switch userEvent {
        case is TaskListEvent.AddTask:
            let event = userEvent as! TaskListEvent.AddTask
            addTask(name: event.name, desc: event.desc)
        case is TaskListEvent.OnClick: doNothing()
        case is TaskListEvent.OnLongClick: doNothing()
        case is TaskListEvent.OnSwipeLeft: doNothing()
        case is TaskListEvent.OnSwipeRight:
            let event = userEvent as! TaskListEvent.OnSwipeRight
            deleteTask(name: event.selectedTask.name)
        case is TaskListEvent.OnRemoveHeadMessageFromQueue: removeMessageFromQueue()
        default: doNothing()
        }
    }
    
    private func loadTasks() {
        loadTasksUseCase.execute().collectCommon(coroutineScope: nil, callback: { output in
            switch output {
            case is LoadTasksOutput.Loading:
                let outputLoading = (output as! LoadTasksOutput.Loading).loading
                self.updateState(loading: outputLoading)
            case is LoadTasksOutput.Data:
                let outputTasks = (output as! LoadTasksOutput.Data).tasks
                self.updateState(taskList: outputTasks)
            case is LoadTasksOutput.Error:
                let outputError = (output as! LoadTasksOutput.Error).message
                self.handleMessage(outputError.build())
            default: self.doNothing()
            }
        })
    }
    
    private func addTask(name: String, desc: String) {
        if let resultMessage = addTaskUseCase.execute(taskName: name, taskDesc: desc) {
            handleMessage(resultMessage.build())
        }
        loadTasks()
    }
    
    private func deleteTask(name: String) {
        if let resultMessage = deleteTaskUseCase.execute(taskName: name) {
            handleMessage(resultMessage.build())
        }
        loadTasks()
    }
    
    private func updateState(
        loading: Bool? = nil,
        taskList: [Task]? = nil,
        queue: Queue<GenericMessageInfo>? = nil
    ) {
        let currentState = (self.state.copy() as! TaskListState)
        self.state = self.state.doCopy(
            isLoading: loading ?? currentState.isLoading,
            tasks: taskList ?? currentState.tasks,
            queue: queue ?? currentState.queue
        )
    }
    
    private func handleMessage(_ message: GenericMessageInfo) {
        switch message.uiComponentType {
        case UIComponentType.Dialog(): appendMessageToQueue(message)
        case UIComponentType.SnackBar(): doNothing()
        case UIComponentType.None(): logger.log(msg: "\(message.description)")
        default: doNothing()
        }
    }
    
    private func appendMessageToQueue(_ message: GenericMessageInfo) {
        let currentState = (self.state.copy() as! TaskListState)
        let queue = currentState.queue
        let queueUtil = GenericMessageInfoQueueUtil()
        if !queueUtil.doesMessageAlreadyExistInQueue(queue: queue, messageInfo: message) {
            queue.add(element: message)
            updateState(queue: queue)
        }
    }
    
    private func removeMessageFromQueue() {
        let queue = (self.state.copy() as! TaskListState).queue
        do {
            try queue.remove()
            updateState(queue: queue)
        } catch {
            self.logger.log(msg: "\(error)")
        }
    }
    
}
